zcash_primitives/src/transaction/components/amount.rs

```diff
@@ -0,0 +1,179 @@
+use std::iter::Sum;
+use std::ops::{Add, AddAssign, Sub, SubAssign};
+
+const COIN: i64 = 1_0000_0000;
+const MAX_MONEY: i64 = 21_000_000 * COIN;
+
+pub const DEFAULT_FEE: Amount = Amount(10000);
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct Amount(i64);
+
+impl Amount {
+    /// Returns a zero-valued Amount.
+    pub const fn zero() -> Self {
+        Amount(0)
+    }
+
+    /// Creates an Amount from an i64.
+    ///
+    /// Returns an error if the amount is out of range.
+    pub fn from_i64(amount: i64, allow_negative: bool) -> Result<Self, ()> {
+        if 0 <= amount && amount <= MAX_MONEY {
+            Ok(Amount(amount))
+        } else if allow_negative && -MAX_MONEY <= amount && amount < 0 {
+            Ok(Amount(amount))
+        } else {
+            Err(())
+        }
+    }
+
+    /// Creates an Amount from a u64.
+    ///
+    /// Returns an error if the amount is out of range.
+    pub fn from_u64(amount: u64) -> Result<Self, ()> {
+        if amount <= MAX_MONEY as u64 {
+            Ok(Amount(amount as i64))
+        } else {
+            Err(())
+        }
+    }
+
+    /// Reads an Amount from a signed 64-bit little-endian integer.
+    ///
+    /// Returns an error if the amount is out of range.
+    pub fn from_i64_le_bytes(bytes: [u8; 8], allow_negative: bool) -> Result<Self, ()> {
+        let amount = i64::from_le_bytes(bytes);
+        Amount::from_i64(amount, allow_negative)
+    }
+
+    /// Reads an Amount from an unsigned 64-bit little-endian integer.
+    ///
+    /// Returns an error if the amount is out of range.
+    pub fn from_u64_le_bytes(bytes: [u8; 8]) -> Result<Self, ()> {
+        let amount = u64::from_le_bytes(bytes);
+        Amount::from_u64(amount)
+    }
+
+    /// Returns the Amount encoded as a signed 64-bit little-endian integer.
+    pub fn to_i64_le_bytes(self) -> [u8; 8] {
+        self.0.to_le_bytes()
+    }
+
+    /// Returns `true` if `self` is positive and `false` if the number is zero or
+    /// negative.
+    pub const fn is_positive(self) -> bool {
+        self.0.is_positive()
+    }
+
+    /// Returns `true` if `self` is negative and `false` if the number is zero or
+    /// positive.
+    pub const fn is_negative(self) -> bool {
+        self.0.is_negative()
+    }
+}
+
+impl From<Amount> for i64 {
+    fn from(amount: Amount) -> i64 {
+        amount.0
+    }
+}
+
+impl From<Amount> for u64 {
+    fn from(amount: Amount) -> u64 {
+        amount.0 as u64
+    }
+}
+
+impl Add<Amount> for Amount {
+    type Output = Amount;
+
+    fn add(self, rhs: Amount) -> Amount {
+        Amount(self.0 + rhs.0)
```

Comment:
Related to the invariant comment, what happens if I create two `Amount`s with `MAX_MONEY - 1` and then add them?  Is that a concern for this struct or is that the user's responsibility?

Reply:
IMHO that's the user's responsibility, from the perspective of this struct. The `Amount`s are either being constructed from visible fields in a transaction (in which case the consensus rules will reject the result, because at least one field contain the sum that is above `MAX_MONEY`, or they are being constructed from/for `Note`s (in which case we are relying on the shielded value balance enforced by the Pedersen commitments, and the commitments themselves being enforced by the circuits). Per my reply to the invariant comment, we could `panic!()` here.